### ISSUE TITLE
Build failure: OpenUSD

### ISSUE THREAD

**Issue by cig0** [NixOS/nixpkgs#325663]
### Steps To Reproduce
1. Try to install the package `python3.11-openusd-24.05`
This is a dependency of some other package, but I don't know which one.

### Expected behavior
The package should install as usual.

### Additional context
```console
error:
       … while calling the 'head' builtin

         at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/attrsets.nix:1575:11:

         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'

         at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/modules.nix:809:9:

          808|     in warnDeprecation opt //
          809|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          810|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Python version mismatch in 'python3.11-openusd-24.05':

       The Python derivation 'python3.11-openusd-24.05' depends on a Python derivation
       named 'python3.12-materialx-1.38.10', but the two derivations use different versions
       of Python:

               'python3.11-openusd-24.05' uses /nix/store/hyppxdhgqx7nbg394j97yhmi6vizl074-python3-3.11.9
           'python3.12-materialx-1.38.10' uses /nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4

       Possible solutions:

         * If 'python3.12-materialx-1.38.10' is a Python library, change the reference to 'python3.12-materialx-1.38.10'
           in the buildInputs of 'python3.11-openusd-24.05' to use a 'python3.12-materialx-1.38.10' built from the same
           version of Python

         * If 'python3.12-materialx-1.38.10' is used as a tool during the build, move the reference to
           'python3.12-materialx-1.38.10' in 'python3.11-openusd-24.05' from buildInputs to nativeBuildInputs

         * If 'python3.12-materialx-1.38.10' provides executables that are called at run time, pass its
           bin path to makeWrapperArgs:

               makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ materialx ] }" ];

        at /nix/store/yzkx7gx2cn73dh3c7zg5rvgkp3yli1cd-source/pkgs/development/python-modules/openusd/default.nix:53:3
┏━ 1 Errors: 
┃ error:
┃        … while calling the 'head' builtin
┃ 
┃          at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/attrsets.nix:1575:11:
┃ 
┃          1574|         || pred here (elemAt values 1) (head values) then
┃          1575|           head values
┃              |           ^
┃          1576|         else
┃ 
┃        … while evaluating the attribute 'value'
┃ 
┃          at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/modules.nix:809:9:
┃ 
┃           808|     in warnDeprecation opt //
┃           809|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
┃              |         ^
┃           810|         inherit (res.defsFinal') highestPrio;
┃ 
┃        (stack trace truncated; use '--show-trace' to show the full trace)
┃ 
┃        error: Python version mismatch in 'python3.11-openusd-24.05':
┃ 
┃        The Python derivation 'python3.11-openusd-24.05' depends on a Python derivation
┃        named 'python3.12-materialx-1.38.10', but the two derivations use different versions
┃        of Python:
┃ 
┃                'python3.11-openusd-24.05' uses /nix/store/hyppxdhgqx7nbg394j97yhmi6vizl074-python3-3.11.9
┃            'python3.12-materialx-1.38.10' uses /nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4
┃ 
┃        Possible solutions:
┃ 
┃          * If 'python3.12-materialx-1.38.10' is a Python library, change the reference to 'python3.12-materialx-1.38.10'
┃            in the buildInputs of 'python3.11-openusd-24.05' to use a 'python3.12-materialx-1.38.10' built from the same
┃            version of Python
┃ 
┃          * If 'python3.12-materialx-1.38.10' is used as a tool during the build, move the reference to
┃            'python3.12-materialx-1.38.10' in 'python3.11-openusd-24.05' from buildInputs to nativeBuildInputs
┃ 
┃          * If 'python3.12-materialx-1.38.10' provides executables that are called at run time, pass its
┃            bin path to makeWrapperArgs:
┃ 
┃                makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ materialx ] }" ];
┃ 
┃         at /nix/store/yzkx7gx2cn73dh3c7zg5rvgkp3yli1cd-source/pkgs/development/python-modules/openusd/default.nix:53:3
┣━━━                                                             
┗━ ∑ ⚠ Exited with 1 errors reported by nix at 19:15:24 after 20s
```

### Notify maintainers
@ShaddyDC  

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.9.7-xanmod1, NixOS, 24.05 (Uakari), 24.05.20240706.49ee0e9`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.4`
 - nixpkgs: `/nix/store/as2lprmykdsqs41yijzdb19asn6iv4g9-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Anifyuli** [comment]
I get this problem too, but it installed as dependency from other derivation. This is my screenshot.

![image](https://github.com/NixOS/nixpkgs/assets/73734843/fc72a666-8933-4e9a-a306-9b50a28e6fb7)

I wish it can fixed soon as possible.

**Comment by returnofblank** [comment]
I did a little searching and I think the issue comes from the `blender` package. or `blender-hip` for me. Removing it from my packages got rid of that error, but now I'm getting a new one.

```
building the system configuration...
error:
       … while calling the 'head' builtin

         at /nix/store/yzkx7gx2cn73dh3c7zg5rvgkp3yli1cd-source/lib/attrsets.nix:1575:11:

         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'

         at /nix/store/yzkx7gx2cn73dh3c7zg5rvgkp3yli1cd-source/lib/modules.nix:821:9:

          820|     in warnDeprecation opt //
          821|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          822|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: sip-4.19.25 not supported for interpreter python3.12

```

**Comment by cig0** [comment]
I removed Blender to give it a try. _The plot thickens!_ (ominous vibe organ playing in the background):
```
error:
       … while calling the 'head' builtin

         at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/attrsets.nix:1575:11:

         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'

         at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/modules.nix:809:9:

          808|     in warnDeprecation opt //
          809|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          810|         inherit (res.defsFinal') highestPrio;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: sip-4.19.25 not supported for interpreter python3.12
┏━ 1 Errors: 
┃ error:
┃        … while calling the 'head' builtin
┃ 
┃          at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/attrsets.nix:1575:11:
┃ 
┃          1574|         || pred here (elemAt values 1) (head values) then
┃          1575|           head values
┃              |           ^
┃          1576|         else
┃ 
┃        … while evaluating the attribute 'value'
┃ 
┃          at /nix/store/2nnisw4pxbifisbkg58hrnis9ycs5ah1-source/lib/modules.nix:809:9:
┃ 
┃           808|     in warnDeprecation opt //
┃           809|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
┃              |         ^
┃           810|         inherit (res.defsFinal') highestPrio;
┃ 
┃        (stack trace truncated; use '--show-trace' to show the full trace)
┃ 
┃        error: sip-4.19.25 not supported for interpreter python3.12
┣━━━                                                             
┗━ ∑ ⚠ Exited with 1 errors reported by nix at 21:51:51 after 31s
Error: 
   0: Command exited with status Exited(1)
```

**Comment by returnofblank** [comment]
Issue could be related to this https://github.com/NixOS/nixpkgs/pull/325682

Either way,
Ma, someone messed with the Python again!

**Comment by ShaddyDC** [comment]
Both plain openusd and blender work for me on master (`nix build .#openusd`, `nix build .#python3Packages.openusd` and `nix build .#blender`).

It's possible that this requires https://github.com/NixOS/nixpkgs/pull/325292 to work properly, which hasn't [propagated](https://nixpk.gs/pr-tracker.html?pr=325292) to nixos-unstable yet?

The python 3.11 version does indeed fail:
```
› nix build .#python311Packages.openusd -L --cores 10
error:
       … while evaluating the attribute 'drvPath'

         at /nix/store/idkl6f3jq1ah9xrznr02yqfbcymxkqp0-source/lib/customisation.nix:365:7:

          364|     in commonAttrs // {
          365|       drvPath = assert condition; drv.drvPath;
             |       ^
          366|       outPath = assert condition; drv.outPath;

       … while evaluating the attribute 'drvPath'

         at /nix/store/idkl6f3jq1ah9xrznr02yqfbcymxkqp0-source/lib/customisation.nix:365:7:

          364|     in commonAttrs // {
          365|       drvPath = assert condition; drv.drvPath;
             |       ^
          366|       outPath = assert condition; drv.outPath;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Python version mismatch in 'python3.11-openusd-24.05':

       The Python derivation 'python3.11-openusd-24.05' depends on a Python derivation
       named 'python3.12-materialx-1.38.10', but the two derivations use different versions
       of Python:

               'python3.11-openusd-24.05' uses /nix/store/hyppxdhgqx7nbg394j97yhmi6vizl074-python3-3.11.9
           'python3.12-materialx-1.38.10' uses /nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4

       Possible solutions:

         * If 'python3.12-materialx-1.38.10' is a Python library, change the reference to 'python3.12-materialx-1.38.10'
           in the buildInputs of 'python3.11-openusd-24.05' to use a 'python3.12-materialx-1.38.10' built from the same
           version of Python

         * If 'python3.12-materialx-1.38.10' is used as a tool during the build, move the reference to
           'python3.12-materialx-1.38.10' in 'python3.11-openusd-24.05' from buildInputs to nativeBuildInputs

         * If 'python3.12-materialx-1.38.10' provides executables that are called at run time, pass its
           bin path to makeWrapperArgs:

               makeWrapperArgs = [ "--prefix PATH : ${lib.makeBinPath [ materialx ] }" ];

        at /nix/store/idkl6f3jq1ah9xrznr02yqfbcymxkqp0-source/pkgs/development/python-modules/openusd/default.nix:53:3
```

I think I'll leave that to people who understand how python derivations properly work here.

**Comment by returnofblank** [comment]
Digressing a little bit, but my earlier error message with sip-4.19.25 was caused by the `cura` package. Worth opening a new issue for?

**Comment by ShaddyDC** [comment]
Probably. Sip4 has explicitly disabled python 3.12, though I don't know why: https://github.com/NixOS/nixpkgs/commit/544c197eb80505155ceea67aac87af9f217f7310
There's some related discussion for cura in https://github.com/NixOS/nixpkgs/issues/186570

It looks like both the packaged version of cura and sip4 are pretty old. Maybe it's worth removing them from whatever depends on them if that's simple.

**Comment by l0b0** [comment]
Also affects hplip.

**Comment by SuperSandro2000** [comment]
> Also affects hplip.

That got fixed in #325682

**Comment by SuperSandro2000** [comment]
> though I don't know why

See the comment: ``relies on distutils``

distutils was dropped in Python 3.12.
